libyul/optimiser/BlockClassFinder.cpp

```diff
@@ -0,0 +1,231 @@
+	hash(_switch.cases.size());
+	ASTWalker::operator()(_switch);
+}
+
+void BlockClassFinder::operator()(FunctionDefinition const& _funDef)
+{
+	hash(literalHash("FunctionDefinition"));
+	m_functionName = _funDef.name;
+	ASTWalker::operator()(_funDef);
+}
+
+void BlockClassFinder::operator()(ForLoop const& _loop)
+{
+	hash(literalHash("ForLoop"));
+	ASTWalker::operator()(_loop);
+}
+
+void BlockClassFinder::operator()(Break const& _break)
+{
+	hash(literalHash("Break"));
+	ASTWalker::operator()(_break);
+}
+
+void BlockClassFinder::operator()(Continue const& _continue)
+{
+	hash(literalHash("Break"));
+	ASTWalker::operator()(_continue);
+}
+
+
+void BlockClassFinder::operator()(Block const& _block)
+{
+	hash(literalHash("Block"));
+	BlockClassFinder subBlockClassFinder(m_globalState);
+	for (auto const& statement: _block.statements)
+		subBlockClassFinder.visit(statement);
+
+	hash(subBlockClassFinder.m_hash);
+	for (auto const& externalIdentifier: subBlockClassFinder.m_externalIdentifiers)
+		(*this)(Identifier{{}, externalIdentifier});
+	for (auto const& externalAssignment: subBlockClassFinder.m_externalAssignments)
+		if (isExternal(externalAssignment))
+			m_externalAssignments.insert(externalAssignment);
+	for (auto const& externalAssignment: subBlockClassFinder.m_externalReads)
+		if (isExternal(externalAssignment))
+			m_externalReads.insert(externalAssignment);
+
+	auto& candidateIDs = m_globalState.hashToBlockIDs[subBlockClassFinder.m_hash];
+
+	// look for existing block class
+	for (auto const& candidateID: candidateIDs)
+	{
+		auto const& candidate = m_globalState.block(candidateID);
+		if (subBlockClassFinder.m_externalIdentifiers.size() == candidate.externalReferences.size())
+		{
+			if (
+				SyntacticallyEqual{
+					subBlockClassFinder.m_externalIdentifiers,
+					candidate.externalReferences
+				}.statementEqual(_block, *candidate.block)
+			)
+			{
+				m_globalState.blockToClassID[&_block] = candidateID.blockClass;
+				m_globalState.blockClasses[candidateID.blockClass].members.emplace_back(BlockClassMember{
+					&_block,
+					std::move(subBlockClassFinder.m_externalIdentifiers),
+					std::move(subBlockClassFinder.m_externalAssignments),
+					std::move(subBlockClassFinder.m_externalReads)
+				});
+				if (!m_functionName.empty())
+				{
+					m_globalState.blockClasses[candidateID.blockClass].nameHint = m_functionName;
+					m_functionName = {};
+				}
+				return;
+			}
+		}
+	}
+
+	// create new block class
+	candidateIDs.emplace_back(GlobalState::BlockID{m_globalState.blockClasses.size(), 0});
+	m_globalState.blockToClassID[&_block] = m_globalState.blockClasses.size();
+	m_globalState.blockClasses.emplace_back(BlockClass{
+		make_vector<BlockClassMember>(
+		std::forward_as_tuple(
+			&_block,
+			std::move(subBlockClassFinder.m_externalIdentifiers),
+			std::move(subBlockClassFinder.m_externalAssignments),
+			std::move(subBlockClassFinder.m_externalReads)
+		)
+	), m_functionName});
+	m_functionName = {};
```

Comment:
This reset (and the one above) result in the function name to be used as name hint only if the block is exactly the body of a function - we might want to use the function name as a hint in general, though - not sure.

Reply:
Function names in code gen tend to get quite long. What I meant with my other comment is that there should be some prefix to hint that it is a function instead of a variable.

Reply:
Ok - I added ``outlined$<start_location>$`` as prefix for regular blocks. I'd still retain the function name, if the body itself is outlined - that will happen for equivalent functions - their body will be outlined entirely, so the original function will be inlined later and we should keep the original function name in the process.
(this can replace equivalent function detection in the end)

Reply:
At least if we want to and I don't see a reason why not - We could also keep equivalent function detection and generally not outline entire function bodies, but equivalent function detection is actually less efficient and outlining all equivalent function bodies plus one step of inlining should be exactly equivalent function detection.